### ISSUE TITLE
Grpc.Core.Internal.CoreErrorDetailException: {"created":"@1621236204.189000000","description":"Received message larger than max (423184132 vs. 67108864)"

### ISSUE THREAD

**Issue by aukevin** [googleads/google-ads-dotnet#285]
I got an interesting error when making a `SearchStream` request to get some responsive search ads:

```
SELECT ad_group_ad.ad.final_urls, ad_group_ad.ad.final_mobile_urls, ad_group_ad.ad.id, ad_group_ad.ad.responsive_search_ad.descriptions, ad_group_ad.ad.responsive_search_ad.headlines, ad_group_ad.ad.responsive_search_ad.path1, ad_group_ad.ad.responsive_search_ad.path2, ad_group_ad.ad.tracking_url_template, ad_group_ad.ad.url_custom_parameters, ad_group_ad.ad.final_url_suffix, ad_group_ad.status, ad_group.id, ad_group.name, campaign.name
FROM ad_group_ad
WHERE campaign.id IN ([redacted])
AND ad_group_ad.ad.type = 'RESPONSIVE_SEARCH_AD'
AND ad_group.status != 'REMOVED'
AND ad_group_ad.status != 'REMOVED'
ORDER BY ad_group_ad.ad.id ASC
```

The error message has pasted below:

```
(Status(StatusCode="ResourceExhausted", Detail="Received message larger than max (423184132 vs. 67108864)", DebugException="Grpc.Core.Internal.CoreErrorDetailException: {"created":"@1621236204.189000000","description":"Received message larger than max (423184132 vs. 67108864)","file":"..\..\..\src\core\ext\filters\message_size\message_size_filter.cc","file_line":208,"grpc_status":8}"))
 ---> Grpc.Core.RpcException: Status(StatusCode="ResourceExhausted", Detail="Received message larger than max (423184132 vs. 67108864)", DebugException="Grpc.Core.Internal.CoreErrorDetailException: {"created":"@1621236204.189000000","description":"Received message larger than max (423184132 vs. 67108864)","file":"..\..\..\src\core\ext\filters\message_size\message_size_filter.cc","file_line":208,"grpc_status":8}")
   at Google.Ads.GoogleAds.Interceptors.StreamingRpcInterceptor`1.MoveNext(CancellationToken cancellationToken)
...
```

It seems like an library-related issue - my understanding was that `SearchStream` handles the message sizes automatically.
Has anyone encountered this issue and/or found a work around?

Thanks,
Kevin

**Comment by aukevin** [comment]
Forgot to include this - this issue occurs on the 6.1.0 version of the Nuget package (which I am working on upgrading) and the V6 Client. I do not know if it has since been fixed.

**Comment by AnashOommen** [comment]
Hi Kevin,

The library limits the maximum response size of a message to 64MB here: https://github.com/googleads/google-ads-dotnet/blob/master/src/Lib/CachedChannelFactory.cs#L42. Your response object is > 400 MB. 

I can try to increase this limit, but just wondering, is there a set of fields you can remove from this query to bring down the message size?

**Comment by aukevin** [comment]
Hi Anash,

Unfortunately, I need all the fields at one point or another to generate the artifact that's being asked of me - it is unlikely that the requirements will be waived.
I can try to split it up, making one set of requests for some fields and other sets of requests for other fields but correlating them costs some overhead (e.g. repeatedly receiving Ad Group Id) that I would like to avoid.

I want to understand this issue better though.
`SearchStream` calls the provided callback in on each `SearchGoogleAdsStreamResponse` it receives.
Is it true that a single response is >64MB?

The reason I ask is because I thought `SearchStream` contrasted with `Search` in how it handled paging.
In particular, instead of having to worry about asking for too many entities in one page and receiving a similar error, I thought `SearchStream` would handle request sizing automatically to not run into such an issue.

Could you explain more about how this works?
Also, any other recommendations to handle this would be appreciated.

Thanks,
Kevin

**Comment by AnashOommen** [comment]
`SearchStream` method uses a fixed 10K `GoogleAdsRow` objects per `SearchGoogleAdsStreamResponse` object, and uses a single stream. The goal is to get the response as fast as possible.

`SearchGoogleAdsRequest` provides you the flexibility to specify the page size, thus controlling the number of `GoogleAdsRow` objects in each `SearchGoogleAdsResponse` object, and each subsequent page is retrieved using a fresh unary API call. The goal is to give you flexibility on how you retrieve the results, at the cost of reduced performance (due to increased number of back-and-forth from your server to Google Servers.

From the API surface, both `Search` and `SearchStream` are similar, so one temporary workaround you can do is to switch this one particular report to use `Search` instead of `SearchStream` and use a page size of 1000 (So your payload size is reduced to 400MB => 40 MB approximately).

I'll talk to the broader team and figure out a way to address this issue. The main open question is how large we should set this parameter to, and whether we should allow the users to override these settings (pros: Additional flexibility, cons: Additional possibility of misconfiguration)
*   grpc supports upto 2GB IIRC (due to the underlying limits on the protobuf library), but increasing the message size tends to have a detrimental effect on the .NET runtime (e.g. https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/large-object-heap is a consideration).
*   grpc itself recommends a default size of 4MB (https://docs.microsoft.com/en-us/aspnet/core/grpc/security?view=aspnetcore-5.0#message-size-limits) for performance reasons, but for the .NET library, we have tested this and tuned to a limit of 64MB.

**Comment by aukevin** [comment]
Hi Anash,

It is good to know the `SearchStream` uses a fixed 10K "page_size", but I have some concerns about it as well.

Please take a look at this (originally) 2019 thread that is related:
https://groups.google.com/g/adwords-api/c/WEOx3jFuGlI/m/lZ28DJHfBAAJ
You can see that support recommended the use of `SearchStream` but, evidently, it does not solve the core issue.
The core issue is that if we must pick a fixed number of entities to get per request, it is always possible that at some point we exceed the size limit.

I am happy to hear that this topic will be under discussion - increasing the limits may help.
To provide my perspective as a user, I would ideally want this "page_size" business to be completely handled by the library.
I could imagine that a request is sent to the server without a "page_size" declared and the server returns as many entities as will fit in a message back until all entities are returned.
However, I understand that there could be limitations preventing this.

**Comment by AnashOommen** [comment]
Hi Kevin,

Let me clarify: Your suggestion is almost how the SearchStream API works today.

1) There's no page size being sent to the server for `SearchGoogleAdsStreamRequest`. See https://github.com/googleapis/googleapis/blob/master/google/ads/googleads/v7/services/google_ads_service.proto#L434

2) For SearchStream API, the server assumes a fixed page size of 10K -- so what you get back is a stream of SearchGoogleAdsStreamResponse objects, each object containing 10K rows (except for the last object, which might have <10K rows). I don't think it is feasible for the server to dynamically adjust the number of rows in the response to stay under 64MB of response, but I can check if such a change is feasible. Is there a request_id I can get for a failed request? It should be in the client library detailed logs.

3) Increasing the `grpc.max_receive_message_length` will definitely address the issue. My suggestion from the previous email was as a temporary workaround.

**Comment by aukevin** [comment]
Hi Anash,

1. Thanks for the link - I understand.

2. Interesting that you mention the request_id. I actually had a separate thread about getting the request id for `SearchStream` requests: https://github.com/googleads/google-ads-dotnet/issues/283. To reduce the volume of logs we have been getting I have it set to only get Information-level request logs. For reasons I am still not sure of, it is inconsistent. Much of the time the RequestId is not present in the log.

@AnashOommen Do you know what level of logging I should set to always get the Request Ids back?
I will try to increase the logging for a short period in the coming days.

3. Sounds good to me!

**Comment by aukevin** [comment]
@AnashOommen Choosing a page size can prove pretty difficult. In addition to the 400MB one from this post, I have recently encountered other sizes (e.g. 160MB). Splitting into pages of 1000 may work well for one instance but be pretty terrible for other instances. Looking forward to any suggestions or improvements to help in these scenarios.

**Comment by AnashOommen** [comment]
I have a release planned in a couple of weeks and can give you a flag to
change this appropriately. Can this wait till then?

**Comment by aukevin** [comment]
Thanks for helping to address this issue! I am fine with waiting and glad to have the confidence that improvements are coming.

**Comment by AnashOommen** [comment]
Hi @aukevin,

I released v9.0.0 now, you can specify something like:

`
config.MaxReceiveMessageLengthInBytes = 2^28;
`

or specify this through your App.config / config.json / alternate configuration mechanisms.

Let me know if this addresses the issue. Note that 2^31 is the maximum that grpc protocol itself supports.

**Comment by aukevin** [comment]
Hi,

Was waiting for some time to make sure the issue didn't occur again.
It looks like adding this feature did the trick!

Thanks,
Kevin
